### ISSUE TITLE
Deprecate `qiskit.extensions` module

### ISSUE THREAD

**Issue by mtreinish** [Qiskit/qiskit-terra#10646]
### What should we add?

The `qiskit.extensions` module is a historical artifact that predates the existence of the circuit library. As we're stabilizing and unifying interface we should remove this standalone module and migrate or remove the classes defined in it. Right now there are the following things in the module:

* `UnitaryGate`
* `Initialize`
* `Snapshot`
* `HamiltonianGate`
* `SingleQubitUnitary`
* `UCPauliRotGate`
* `UCRXGate`
* `UCRYGate`
* `UCRYGate`
* `Isometry`
* `DiagonalGate`
* `MCGupDiag`
* `UCGate`

As a first step we should migrate these class definitions to live in the circuit library and add a redirect shim that will emit a deprecation warning for anything using `qiskit.extensions`. In parallel we should take inventory of these classes and individually deprecate any class which doesn't provide value anymore and stage those for removal.

**Comment by mtreinish** [comment]
This is a duplicate of: https://github.com/Qiskit/qiskit-terra/issues/6861
